File: src/main/scala/otserver4s/database/Account.scala
package otserver4s.database

import com.j256.ormlite.field.DatabaseField
import com.j256.ormlite.support.ConnectionSource
import org.apache.log4j.Logger
import otserver4s.OTServerLoginException
import otserver4s.login.LoginRequest
import otserver4s.utils.MD5Utils

@com.j256.ormlite.table.DatabaseTable(tableName = "contas")
class Conta {

  @DatabaseField(id = true, unique = true)
  var number: Integer = null

  @DatabaseField(canBeNull = false) 
  var hashSenha: String = null

  @DatabaseField(canBeNull = false) 
  var diasPremiumRestantes: Integer = null

  var personagens: List[Personagem] = null

  def adicionarPersonagens(personagens: List[Personagem]) = {
    this.personagens = personagens
    this
  }
  
  override def toString = 
    s"Conta(number: $number, hashSenha: $hashSenha, " +
    s"premium: $diasPremiumRestantes" +
    (if(personagens != null) s", personagens: ${personagens.length}" 
     else "") + ")" 
  
}

object Conta {
  def criarTabelaSeNaoExistir(conexao: ConnectionSource) =
    com.j256.ormlite.table.TableUtils.createTableIfNotExists(
      conexao, classOf[Conta])
  private def criarDAO(conexao: ConnectionSource):
    com.j256.ormlite.dao.Dao[Conta, Integer] = 
      com.j256.ormlite.dao.DaoManager.createDao(conexao, classOf[Conta])
  def contarTodos(conexao: ConnectionSource) = 
    criarDAO(conexao).countOf
  def persistirNova(number: Integer, senha: String, conexao: ConnectionSource) = 
    (Option(number), Option(senha)) match {
      case (Some(accNumber), Some(senhaDescriptografada)) => {
        val account = new Conta
        account.number = accNumber
        account.hashSenha = MD5Utils.str2md5(senhaDescriptografada)
        account.diasPremiumRestantes = 0
        criarDAO(conexao).create(account)
      }
      case (_, _) => throw new OTServerLoginException("mensagem.login.salvar.invalido")
    }
  def apply(loginRequest: LoginRequest, conexao: ConnectionSource, 
      carregarPersonagens: Boolean = true): Conta =
    loginRequest.versao match {
      case LoginRequest.VERSAO => {
        (loginRequest.accountNumber, loginRequest.password) match {
          case (None, None) => throw new OTServerLoginException("mensagem.login.conta.senha.vazia")
          case (None, Some(y)) => throw new OTServerLoginException("mensagem.login.conta.vazia")
          case (Some(x), None) => throw new OTServerLoginException("mensagem.login.senha.vazia")
          case (Some(x), Some(y)) => {
            val dao = criarDAO(conexao)
            Option(dao.queryForId(loginRequest.accountNumber.get)) match {
              case None => throw new OTServerLoginException("mensagem.login.conta.inexistente")
              case Some(account) =>
                if(loginRequest.password.get == account.hashSenha) 
                  if(carregarPersonagens)account.adicionarPersonagens(
                    Personagem.listarPersonagensPorAccount(account.number, conexao))
                  else account
                else throw new OTServerLoginException("mensagem.login.senha.incorreta")
            }
          }
        }
      }
      case _ => throw new OTServerLoginException("mensagem.login.versao.errada")
    }
  
}

File: src/main/kotlin/otkserver/Account.kt
package otkserver

import com.j256.ormlite.field.DatabaseField
import com.j256.ormlite.field.ForeignCollectionField
import com.j256.ormlite.support.ConnectionSource
import com.j256.ormlite.dao.ForeignCollection

@com.j256.ormlite.table.DatabaseTable(tableName = "contas")
class Conta(
  @DatabaseField(id = true, unique = true)
  var codigo: Int = 0,
  
  @DatabaseField(canBeNull = false) 
  var hashSenha: String = "",
  
  @DatabaseField(canBeNull = false) 
  var diasPremiumRestantes: Int = 0,
  
  @ForeignCollectionField(eager = false)
  var personagens: ForeignCollection<Personagem>? = null
) {
	override fun toString() =
		"Account(number=$codigo, senha=$hashSenha, premmy=$diasPremiumRestantes)"
	fun recarregar(conexao: ConnectionSource): Conta {
		Conta.criarDAO(conexao).refresh(this)
		return this
	}
	fun salvar(conexao: ConnectionSource) {
		this.hashSenha = MD5Utils.str2md5(this.hashSenha)
    Conta.criarDAO(conexao).create(this)
	}
	companion object {
		fun criarTabelaSeNaoExistir(conexao: ConnectionSource) =
			com.j256.ormlite.table.TableUtils.createTableIfNotExists(
      conexao, Conta::class.java)
		private fun criarDAO(conexao: ConnectionSource):
		  com.j256.ormlite.dao.Dao<Conta, Int> = 
      com.j256.ormlite.dao.DaoManager.createDao(conexao, Conta::class.java)
		fun contarTodos(conexao: ConnectionSource) = criarDAO(conexao).countOf()
		fun buscarPorProtocoloLogin(
			  protocoloLogin: ProtocoloLogin,
			  conexao: ConnectionSource): Conta {
			criarDAO(conexao).queryForId(protocoloLogin.account.number)?.let {
				return if(protocoloLogin.account.password.equals(it.hashSenha)) it
				else throw OTServerLoginException("mensagem.login.senha.incorreta")
			}
			throw OTServerLoginException("mensagem.login.conta.inexistente")
		}
	}
}
